File: app/src/main/java/com/expansion/lg/kimaru/training/adapters/TraineesExamSubmissionListAdapter.java
package com.expansion.lg.kimaru.training.adapters;

import android.content.Context;
import android.support.v4.content.ContextCompat;
import android.support.v7.widget.RecyclerView.Adapter;
import android.support.v7.widget.RecyclerView.ViewHolder;
import android.util.SparseBooleanArray;
import android.view.HapticFeedbackConstants;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.expansion.lg.kimaru.training.R;
import com.expansion.lg.kimaru.training.database.DatabaseHelper;
import com.expansion.lg.kimaru.training.objs.TrainingExam;
import com.expansion.lg.kimaru.training.objs.TrainingExamResult;
import com.expansion.lg.kimaru.training.objs.TrainingTrainee;
import com.expansion.lg.kimaru.training.utils.FlipAnimator;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by kimaru on 3/15/18 using Android Studio.
 * Maintainer: David Kimaru
 *
 * @github https://github.com/kimarudg
 * @email kimarudg@gmail.com
 * @phone +254722384549
 * @web gakuu.co.ke
 */

public class TraineesExamSubmissionListAdapter extends Adapter<TraineesExamSubmissionListAdapter.ListHolder>{
    private Context context;
    private List<TrainingTrainee> trainees;
    private TraineesExamSubmissionListAdapterListener listener;
    private SparseBooleanArray selectedItems, selectedItemsIndex;
    private boolean reverseAllActions = false;
    private static int currentSelectedIndex = -1;
    private TrainingExam exam;
    private int noOfQuestions = 0;
    DatabaseHelper databaseHelper;

    public class ListHolder extends ViewHolder implements View.OnLongClickListener{
        public TextView title, subTitle, message, iconText, timestamp;
        public ImageView iconImp, imgProfile;
        public LinearLayout container;
        public RelativeLayout viewBackground, viewForeground, iconContainer, iconBack, iconFront;

        public ListHolder(View view){
            super(view);

            title = (TextView) view.findViewById(R.id.from);
            subTitle = (TextView) view.findViewById(R.id.txt_primary);
            message = (TextView) view.findViewById(R.id.txt_secondary);
            iconText = (TextView) view.findViewById(R.id.icon_text);
            timestamp = (TextView) view.findViewById(R.id.timestamp);
            iconBack = (RelativeLayout) view.findViewById(R.id.icon_back);
            iconFront = (RelativeLayout) view.findViewById(R.id.icon_front);
            iconImp = (ImageView) view.findViewById(R.id.icon_star);
            imgProfile = (ImageView) view.findViewById(R.id.icon_profile);
            container = (LinearLayout) view.findViewById(R.id.message_container);
            iconContainer = (RelativeLayout) view.findViewById(R.id.icon_container);
            viewBackground = view.findViewById(R.id.view_background);
            viewForeground = view.findViewById(R.id.view_foreground);

            view.setOnLongClickListener(this);
        }

        @Override
        public boolean onLongClick(View view){
            listener.onRowLongClicked(getAdapterPosition());
            view.performHapticFeedback(HapticFeedbackConstants.LONG_PRESS);
            return true;
        }
    }
    public TraineesExamSubmissionListAdapter(Context context, List<TrainingTrainee> trainingTrainees,
                                             TraineesExamSubmissionListAdapterListener listener, TrainingExam exam){
        this.context = context;
        this.trainees = trainingTrainees;
        this.listener = listener;
        selectedItems = new SparseBooleanArray();
        selectedItemsIndex = new SparseBooleanArray();
        this.exam = exam;
        try {
            noOfQuestions = exam.getCloudExamJson().getJSONArray("questions").length();
        }catch (Exception e){}
        databaseHelper = new DatabaseHelper(context);
    }

    @Override
    public ListHolder onCreateViewHolder(ViewGroup parent, int viewType){
        View itemView = LayoutInflater.from(parent.getContext()).inflate(
                R.layout.recycler_list_row, parent, false);
        return new ListHolder(itemView);
    }

    @Override
    public void onBindViewHolder(final ListHolder holder, final int position){
        TrainingTrainee trainee = trainees.get(position);

        //Get all submissions by this trainee for this exam
        List<TrainingExamResult>  submittedQuestions = databaseHelper
                .getTrainingExamResultByExamTraineeId(exam.getId().toString(), trainee.getRegistrationId());

        int noOfSubmittedQuestions = submittedQuestions.size();
        String submissionStatus = "";
        String examStatus = "UNKNOWN";
        String examTotalMarks = "--";
        Integer totalMarks = 0;

        if (noOfSubmittedQuestions == 0){
            submissionStatus = "Undone";
            examStatus = "UNKNOWN";
        }else if (noOfSubmittedQuestions < noOfQuestions){
            submissionStatus = "Incomplete";
            for (TrainingExamResult result : submittedQuestions){
                totalMarks += result.getQuestionScore();
            }
            examTotalMarks = String.format("Marks so far %s", totalMarks.toString());
            examStatus = "CALCULATING ... ";
        }else {
            submissionStatus = "Complete";
            for (TrainingExamResult result : submittedQuestions){
                totalMarks += result.getQuestionScore();
            }
            examTotalMarks = totalMarks.toString();
            examStatus = totalMarks >= exam.getPassmark() ? "PASS" : "FAILED";

            if (exam.getPassmark() == 0) {
                // Trainer did not set a passmark for this exam
                examStatus = "UNKNOWN";
            }
        }
        holder.subTitle.setText(examTotalMarks);

        //@TODO: create configuration for pass status (pass, average, fail, or unknown - in case someone has not submitted the exam)

        holder.message.setText(examStatus);

        holder.timestamp.setText(submissionStatus);
        try{
            holder.title.setText(trainee.getRegistration().getString("name"));
            // displaying the first letter of From in icon text
            holder.iconText.setText(trainee.getRegistration().getString("name").substring(0,1));
            // change the row state to activated
        }catch (Exception e){}
        holder.itemView.setActivated(selectedItems.get(position, false));
        // handle icon animation
        applyIconAnimation(holder, position);
        applyIconColor(holder, trainee);
        // apply click events
        applyClickEvents(holder, position);
        applyImportant(holder, submissionStatus);

    }

    private void applyClickEvents(ListHolder holder, final int position){
        holder.iconContainer.setOnClickListener(new View.OnClickListener(){
            @Override
            public void onClick(View view){
                listener.onIconClicked(position);
            }
        });
        holder.iconImp.setOnClickListener(new View.OnClickListener(){
            @Override
            public void onClick(View view){
                listener.onIconImportantClicked(position);
            }
        });
        holder.container.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                listener.onMessageRowClicked(position);
            }
        });
        holder.container.setOnLongClickListener(new View.OnLongClickListener(){
            @Override
            public boolean onLongClick(View view){
                listener.onRowLongClicked(position);
                view.performHapticFeedback(HapticFeedbackConstants.LONG_PRESS);
                return true;
            }
        });
    }

    private void applyIconAnimation(ListHolder holder, int position) {
        if (selectedItems.get(position, false)) {
            holder.iconFront.setVisibility(View.GONE);
            resetIconYAxis(holder.iconBack);
            holder.iconBack.setVisibility(View.VISIBLE);
            holder.iconBack.setAlpha(1);
            if (currentSelectedIndex == position) {
                FlipAnimator.flipView(context, holder.iconBack, holder.iconFront, true);
                resetCurrentIndex();
            }
        } else {
            holder.iconBack.setVisibility(View.GONE);
            resetIconYAxis(holder.iconFront);
            holder.iconFront.setVisibility(View.VISIBLE);
            holder.iconFront.setAlpha(1);
            if ((reverseAllActions && selectedItemsIndex.get(position, false)) || currentSelectedIndex == position) {
                FlipAnimator.flipView(context, holder.iconBack, holder.iconFront, false);
                resetCurrentIndex();
            }
        }
    }

    private void applyIconColor(ListHolder holder, TrainingTrainee trainingTrainee) {
        holder.imgProfile.setImageResource(R.drawable.bg_circle);
        holder.imgProfile.setColorFilter(trainingTrainee.getColor());
        holder.iconText.setVisibility(View.VISIBLE);
    }


    // As the views will be reused, sometimes the icon appears as
    // flipped because older view is reused. Reset the Y-axis to 0
    private void resetIconYAxis(View view) {
        if (view.getRotationY() != 0) {
            view.setRotationY(0);
        }
    }

    public void resetAnimationIndex() {
        reverseAllActions = false;
        selectedItemsIndex.clear();
    }

    private void applyImportant(ListHolder holder, String submissionStatus) {
        if (submissionStatus.equalsIgnoreCase("Complete")) {
            holder.iconImp.setImageDrawable(ContextCompat.getDrawable(context, R.drawable.ic_completed_24px));
            holder.iconImp.setColorFilter(ContextCompat.getColor(context, R.color.LightGreen));
        } else if (submissionStatus.equalsIgnoreCase("Undone")) {
            holder.iconImp.setImageDrawable(ContextCompat.getDrawable(context, R.drawable.ic_not_done));
            holder.iconImp.setColorFilter(ContextCompat.getColor(context, R.color.Red));
        }else {
            holder.iconImp.setImageDrawable(ContextCompat.getDrawable(context, R.drawable.ic_waiting_24px));
            holder.iconImp.setColorFilter(ContextCompat.getColor(context, R.color.icon_tint_normal));
        }
    }

    @Override
    public int getItemCount(){
        return trainees.size();
    }

    public void toggleSelection(int pos) {
        currentSelectedIndex = pos;
        if (selectedItems.get(pos, false)) {
            selectedItems.delete(pos);
            selectedItemsIndex.delete(pos);
        } else {
            selectedItems.put(pos, true);
            selectedItemsIndex.put(pos, true);
        }
        notifyItemChanged(pos);
    }

    public void clearSelections() {
        reverseAllActions = true;
        selectedItems.clear();
        notifyDataSetChanged();
    }
    public int getSelectedItemCount() {
        return selectedItems.size();
    }

    public List<Integer> getSelectedItems() {
        List<Integer> items =
                new ArrayList<>(selectedItems.size());
        for (int i = 0; i < selectedItems.size(); i++) {
            items.add(selectedItems.keyAt(i));
        }
        return items;
    }

    public void removeData(int position) {
        trainees.remove(position);
        notifyItemRemoved(position);
        resetCurrentIndex();
    }
    public void restoreItem(TrainingTrainee trainingTrainee, int position) {
        trainees.add(position, trainingTrainee);
        // notify item added by position
        notifyItemInserted(position);
    }

    private void resetCurrentIndex() {
        currentSelectedIndex = -1;
    }



    public interface TraineesExamSubmissionListAdapterListener{
        void onIconClicked(int position);
        void onIconImportantClicked(int position);
        void onMessageRowClicked(int position);
        void onRowLongClicked(int position);
    }
}

File: app/src/main/java/com/expansion/lg/kimaru/training/adapters/TraineesExamSubmissionListAdapter.kt
package com.expansion.lg.kimaru.training.adapters

import android.content.Context
import android.support.v4.content.ContextCompat
import android.support.v7.widget.RecyclerView.Adapter
import android.support.v7.widget.RecyclerView.ViewHolder
import android.util.SparseBooleanArray
import android.view.HapticFeedbackConstants
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.RelativeLayout
import android.widget.TextView

import com.expansion.lg.kimaru.training.R
import com.expansion.lg.kimaru.training.database.DatabaseHelper
import com.expansion.lg.kimaru.training.objs.TrainingExam
import com.expansion.lg.kimaru.training.objs.TrainingExamResult
import com.expansion.lg.kimaru.training.objs.TrainingTrainee
import com.expansion.lg.kimaru.training.utils.DisplayDate
import com.expansion.lg.kimaru.training.utils.FlipAnimator

import java.util.ArrayList

/**
 * Created by kimaru on 3/15/18 using Android Studio.
 * Maintainer: David Kimaru
 *
 * @github https://github.com/kimarudg
 * @email kimarudg@gmail.com
 * @phone +254722384549
 * @web gakuu.co.ke
 */

class TraineesExamSubmissionListAdapter(private val context: Context, private val trainees: MutableList<TrainingTrainee>,
                                        private val listener: TraineesExamSubmissionListAdapterListener, private val exam: TrainingExam) : Adapter<TraineesExamSubmissionListAdapter.ListHolder>() {
    private val selectedItems: SparseBooleanArray
    private val selectedItemsIndex: SparseBooleanArray
    private var reverseAllActions = false
    private var noOfQuestions = 0
    internal var databaseHelper: DatabaseHelper
    val selectedItemCount: Int
        get() = selectedItems.size()

    inner class ListHolder(view: View) : ViewHolder(view), View.OnLongClickListener {
        var title: TextView
        var subTitle: TextView
        var message: TextView
        var iconText: TextView
        var timestamp: TextView
        var iconImp: ImageView
        var imgProfile: ImageView
        var container: LinearLayout
        var viewBackground: RelativeLayout
        var viewForeground: RelativeLayout
        var iconContainer: RelativeLayout
        var iconBack: RelativeLayout
        var iconFront: RelativeLayout

        init {

            title = view.findViewById<View>(R.id.from) as TextView
            subTitle = view.findViewById<View>(R.id.txt_primary) as TextView
            message = view.findViewById<View>(R.id.txt_secondary) as TextView
            iconText = view.findViewById<View>(R.id.icon_text) as TextView
            timestamp = view.findViewById<View>(R.id.timestamp) as TextView
            iconBack = view.findViewById<View>(R.id.icon_back) as RelativeLayout
            iconFront = view.findViewById<View>(R.id.icon_front) as RelativeLayout
            iconImp = view.findViewById<View>(R.id.icon_star) as ImageView
            imgProfile = view.findViewById<View>(R.id.icon_profile) as ImageView
            container = view.findViewById<View>(R.id.message_container) as LinearLayout
            iconContainer = view.findViewById<View>(R.id.icon_container) as RelativeLayout
            viewBackground = view.findViewById(R.id.view_background)
            viewForeground = view.findViewById(R.id.view_foreground)

            view.setOnLongClickListener(this)
        }

        override fun onLongClick(view: View): Boolean {
            listener.onRowLongClicked(adapterPosition)
            view.performHapticFeedback(HapticFeedbackConstants.LONG_PRESS)
            return true
        }
    }

    init {
        selectedItems = SparseBooleanArray()
        selectedItemsIndex = SparseBooleanArray()
        try {
            noOfQuestions = exam.getCloudExamJson().getJSONArray("questions").length()
        } catch (e: Exception) {
        }

        databaseHelper = DatabaseHelper(context)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ListHolder {
        val itemView = LayoutInflater.from(parent.context).inflate(
                R.layout.recycler_list_row, parent, false)
        return ListHolder(itemView)
    }

    override fun onBindViewHolder(holder: ListHolder, position: Int) {
        val trainee = trainees[position]

        //Get all submissions by this trainee for this exam
        val submittedQuestions = databaseHelper
                .getTrainingExamResultByTrainee(trainee.registrationId)

        val noOfSubmittedQuestions = submittedQuestions.size
        var submissionStatus = ""
        var examStatus = "UNKNOWN"
        var examTotalMarks = "--"
        var totalMarks: Int? = 0

        if (noOfSubmittedQuestions == 0) {
            submissionStatus = "Undone"
            examStatus = "UNKNOWN"
        } else if (noOfSubmittedQuestions < noOfQuestions) {
            submissionStatus = "Incomplete"
            for (result in submittedQuestions) {
                totalMarks += result.questionScore
            }
            examTotalMarks = String.format("Marks so far %s", totalMarks!!.toString())
            examStatus = "CALCULATING ... "
        } else {
            submissionStatus = "Complete"
            for (result in submittedQuestions) {
                totalMarks += result.questionScore
            }
            examTotalMarks = totalMarks!!.toString()
            examStatus = if (totalMarks >= exam.passmark) "PASS" else "FAILED"
        }
        holder.subTitle.text = examTotalMarks

        //@TODO: create configuration for pass status (pass, average, fail, or unknown - in case someone has not submitted the exam)

        holder.message.text = examStatus

        holder.timestamp.text = submissionStatus
        try {
            holder.title.text = trainee.registration.getString("name")
            // displaying the first letter of From in icon text
            holder.iconText.setText(trainee.registration.getString("name").substring(0, 1))
            // change the row state to activated
        } catch (e: Exception) {
        }

        holder.itemView.isActivated = selectedItems.get(position, false)
        // handle icon animation
        applyIconAnimation(holder, position)
        applyIconColor(holder, trainee)
        // apply click events
        applyClickEvents(holder, position)
        applyImportant(holder, submissionStatus)

    }

    private fun applyClickEvents(holder: ListHolder, position: Int) {
        holder.iconContainer.setOnClickListener { listener.onIconClicked(position) }
        holder.iconImp.setOnClickListener { listener.onIconImportantClicked(position) }
        holder.container.setOnClickListener { listener.onMessageRowClicked(position) }
        holder.container.setOnLongClickListener { view ->
            listener.onRowLongClicked(position)
            view.performHapticFeedback(HapticFeedbackConstants.LONG_PRESS)
            true
        }
    }

    private fun applyIconAnimation(holder: ListHolder, position: Int) {
        if (selectedItems.get(position, false)) {
            holder.iconFront.visibility = View.GONE
            resetIconYAxis(holder.iconBack)
            holder.iconBack.visibility = View.VISIBLE
            holder.iconBack.alpha = 1f
            if (currentSelectedIndex == position) {
                FlipAnimator.flipView(context, holder.iconBack, holder.iconFront, true)
                resetCurrentIndex()
            }
        } else {
            holder.iconBack.visibility = View.GONE
            resetIconYAxis(holder.iconFront)
            holder.iconFront.visibility = View.VISIBLE
            holder.iconFront.alpha = 1f
            if (reverseAllActions && selectedItemsIndex.get(position, false) || currentSelectedIndex == position) {
                FlipAnimator.flipView(context, holder.iconBack, holder.iconFront, false)
                resetCurrentIndex()
            }
        }
    }

    private fun applyIconColor(holder: ListHolder, trainingTrainee: TrainingTrainee) {
        holder.imgProfile.setImageResource(R.drawable.bg_circle)
        holder.imgProfile.setColorFilter(trainingTrainee.color)
        holder.iconText.visibility = View.VISIBLE
    }


    // As the views will be reused, sometimes the icon appears as
    // flipped because older view is reused. Reset the Y-axis to 0
    private fun resetIconYAxis(view: View) {
        if (view.rotationY != 0f) {
            view.rotationY = 0f
        }
    }

    fun resetAnimationIndex() {
        reverseAllActions = false
        selectedItemsIndex.clear()
    }

    private fun applyImportant(holder: ListHolder, submissionStatus: String) {
        if (submissionStatus.equals("Complete", ignoreCase = true)) {
            holder.iconImp.setImageDrawable(ContextCompat.getDrawable(context, R.drawable.ic_completed_24px))
            holder.iconImp.setColorFilter(ContextCompat.getColor(context, R.color.LightGreen))
        } else if (submissionStatus.equals("Undone", ignoreCase = true)) {
            holder.iconImp.setImageDrawable(ContextCompat.getDrawable(context, R.drawable.ic_not_done))
            holder.iconImp.setColorFilter(ContextCompat.getColor(context, R.color.Red))
        } else {
            holder.iconImp.setImageDrawable(ContextCompat.getDrawable(context, R.drawable.ic_waiting_24px))
            holder.iconImp.setColorFilter(ContextCompat.getColor(context, R.color.icon_tint_normal))
        }
    }

    override fun getItemCount(): Int {
        return trainees.size
    }

    fun toggleSelection(pos: Int) {
        currentSelectedIndex = pos
        if (selectedItems.get(pos, false)) {
            selectedItems.delete(pos)
            selectedItemsIndex.delete(pos)
        } else {
            selectedItems.put(pos, true)
            selectedItemsIndex.put(pos, true)
        }
        notifyItemChanged(pos)
    }

    fun clearSelections() {
        reverseAllActions = true
        selectedItems.clear()
        notifyDataSetChanged()
    }

    fun getSelectedItems(): List<Int> {
        val items = ArrayList<Int>(selectedItems.size())
        for (i in 0 until selectedItems.size()) {
            items.add(selectedItems.keyAt(i))
        }
        return items
    }

    fun removeData(position: Int) {
        trainees.removeAt(position)
        notifyItemRemoved(position)
        resetCurrentIndex()
    }

    fun restoreItem(trainingTrainee: TrainingTrainee, position: Int) {
        trainees.add(position, trainingTrainee)
        // notify item added by position
        notifyItemInserted(position)
    }

    private fun resetCurrentIndex() {
        currentSelectedIndex = -1
    }


    interface TraineesExamSubmissionListAdapterListener {
        fun onIconClicked(position: Int)
        fun onIconImportantClicked(position: Int)
        fun onMessageRowClicked(position: Int)
        fun onRowLongClicked(position: Int)
    }

    companion object {
        private var currentSelectedIndex = -1
    }
}
